Feature/regional bucket s3 request

darrell-d opened the pull request:
Undoing `withForceGlobalBucketAccessEnabled` flag since it didn't work

Moving to making a client factory to produce a new client only when it is not in the us-east-1 region

muftring commented:
Hey @darrell-d , a few comments and thoughts...
1. do we need region-specific S3 Clients for the two other functions in the `S3ObjectStore` class?
2. we could simplify `getPresignedUrl()` to always get an S3 Client from the `S3ClientFactory`, even for the us-east-1 region. This would mean we would have two S3 Clients created for us-east-1, the class instance value provided in the constructor and the one in the `S3ClientFactory`. Of course, we could refactor `S3ObjectStore` to eliminate the class instance value in the constructor if the two other methods are refactored to use the `S3ClientFactory`.
3. the `S3ClientFactory` could be simplified to just a single method: (something like) `getClientForBucket()` which would extract the bucket suffix, map it to a region, and yield an S3 Client.
4. the `regionMappings` Map should be a `private val` on the object not a val in the function -- this will save the cost of construction on each invocation though the compiler might already be optimizing this).

darrell-d commented:
> Hey @darrell-d , a few comments and thoughts...
> 
> 1. do we need region-specific S3 Clients for the two other functions in the `S3ObjectStore` class?
> 2. we could simplify `getPresignedUrl()` to always get an S3 Client from the `S3ClientFactory`, even for the us-east-1 region. This would mean we would have two S3 Clients created for us-east-1, the class instance value provided in the constructor and the one in the `S3ClientFactory`. Of course, we could refactor `S3ObjectStore` to eliminate the class instance value in the constructor if the two other methods are refactored to use the `S3ClientFactory`.
> 3. the `S3ClientFactory` could be simplified to just a single method: (something like) `getClientForBucket()` which would extract the bucket suffix, map it to a region, and yield an S3 Client.
> 4. the `regionMappings` Map should be a `private val` on the object not a val in the function -- this will save the cost of construction on each invocation though the compiler might already be optimizing this).

These suggestions make sense. I did want to try to patch the issues quickly, but not too haphazardly.

1/2. I updated the factory to simplify it, and now use the regionalClient on all the functions in S3ObjectStore. Also removed the S3 instance from S3ObjectStores around the codebase
3. Changed it so it's a single call, but I do keep a private method as a helper. I ran into some scala return type issues when I tried to do it inline, vs a function call. Wasn't sure what the difference was.
4. Good catch. Change

muftring reviewed (approved):
looks good
